app/models/concerns/unparsed_run.rb

```diff
@@ -13,154 +18,286 @@ def parse_into_db
-          realtime_sum_of_best_ms: zero_to_nil(parse_result[:realtime_sum_of_best_ms]),
+          realtime_duration_ms:    zero_to_nil(run_data[:realtime_duration_ms]),
+          realtime_sum_of_best_ms: zero_to_nil(run_data[:realtime_sum_of_best_ms]),
 
-          gametime_duration_ms:    zero_to_nil(parse_result[:gametime_duration_ms]),
-          gametime_sum_of_best_ms: zero_to_nil(parse_result[:gametime_sum_of_best_ms]),
+          gametime_duration_ms:    zero_to_nil(run_data[:gametime_duration_ms]),
+          gametime_sum_of_best_ms: zero_to_nil(run_data[:gametime_sum_of_best_ms]),
 
-          total_playtime_ms:       parse_result[:total_playtime_ms],
-          default_timing:          default_timing,
-          filesize_bytes:          size
+          total_playtime_ms:       run_data[:total_playtime_ms],
+          default_timing:          run_data[:default_timing],
+          filesize_bytes:          run_data[:size]
         )
 
         HighlightSuggestion.from_run(self)
       end
+    rescue UnparsableRun
+      destroy
+      false
+    ensure
+      run.dispose if defined?(run)
     end
 
-    def write_run_histories(histories)
-      return nil if histories.nil?
+    def process_run_wrapper(wrapper, run_data)
+      run_data[:program] = wrapper.timer_kind
+      run_data[:program] = Programs::ExchangeFormat.to_s if Run.program_from_attribute(:to_s, run_data[:program]).nil?
+
+      if wrapper.is_generic_timer && !Run.program_from_attribute(:to_s, run_data[:program]).exchangeable?
+        # We got a file in the exchange format, but the reported timer doesn't support the exchange format.
+        # Wipe the timer to stop from serving the exchange format when a user asks to download the file in the
+        # source timer's format.
+        run_data[:program] = Programs::ExchangeFormat.to_s
+      end
+
+      run_data[:program] = Run.program_from_attribute(:to_s, run_data[:program]).to_sym
+
+      run_data
+    end
+
+    def process_run_metadata(run, run_data)
+      run_data[:game]      = run.game_name
+      run_data[:game_icon] = run.game_icon.presence
+      run_data[:category]  = run.category_name
+      run_data[:name]      = run.extended_name(false)
+
+      run_data[:metadata]                 = {}
+      run_data[:metadata][:srdc_id]       = run.metadata.run_id.presence
+      run_data[:metadata][:platform_name] = run.metadata.platform_name.presence
+      run_data[:metadata][:region_name]   = run.metadata.region_name.presence
+      run_data[:metadata][:uses_emulator] = run.metadata.uses_emulator
+      run_data[:metadata][:variables]     = {}
+
+      run_data[:attempts]  = run.attempt_count
+      run_data[:offset_ms] = run.offset.total_seconds * 1000
+
+      # Grab all run metadata
+      # .with will clean up the iterator when we are done
+      run.metadata.variables.with do |iter|
+        until (variable = iter.next).nil?
+          run_data[:metadata][:variables][variable.name] = variable.value
```

Comment:
Variables still aren't being saved, but since it's such a small part of the parser it didn't make much sense to remove it now.